server/oauth.js

```diff
@@ -0,0 +1,299 @@
+  , generateUnique = require('time-uuid')
+  , request        = require('request')
+  , jwtDecode      = require('jwt-decode')
+  , env            = mano.env
+  , userStorage    = mano.dbDriver.getStorage('user');
+
+var dbjsDataRecord = function (id, value) {
+	return { id: id, data: { value: serializeValue(value) } };
+};
+
+var createUser = function (data) {
+	return mano.queryMemoryDb([], 'addUser', JSON.stringify(data));
+};
+
+var updateUser = function (userId, data) {
+	var records = [];
+
+	Object.keys(data).forEach(function (key) {
+		records.push(dbjsDataRecord(userId + '/' + key, data[key]));
+	});
+
+	return userStorage.storeMany(records)(function () {
+		return loadToMemoryDb([userId]);
+	})(function () {
+		return userId;
+	});
+};
+
+var generateUrl = function (path, query) {
+	var keys = Object.keys(query);
+
+	keys.forEach(function (key, index) {
+		if (index === 0) path += '?';
+
+		path += key + '=' + encodeURIComponent(query[key]);
+
+		if (index !== (keys.length - 1)) path += '&';
+	});
+
+	return path;
+};
+
+module.exports = exports = {
+	loginMiddleware: function (req, res, next) {
+		// 1. If not proper path, end.
+		if (req._parsedUrl.pathname !== '/oauth-login/') {
+			next();
+			return;
+		}
+
+		// 2. Is there a user logged in?
+		if (req.$user) {
+			var demoUserId      = res.cookies.get('demoUser')
+			  , isDifferentUser = demoUserId !== req.$user;
+
+			// 2.1. If he is not current demo user, end.
+			if (!demoUserId || isDifferentUser) {
+				next();
+				return;
+			}
+		}
+
+		// 3. Generate and store unique state.
+		var state = generateUnique();
+		res.cookies.set('OAuthState', state);
+
+		debug('Login request. Unique state:', state);
+
+		// 4. Build redirect query.
+		var locationQuery = assign({}, req.query, {
+			response_type: 'code',
+			client_id: env.oauth.clientId,
+			state: state,
+			scope: env.oauth.scope,
+			redirect_uri: env.oauth.redirectUrl
+		});
+
+		if (env.oauth.extraAuthorizationParameters) {
+			assign(locationQuery, env.oauth.extraAuthorizationParameters);
+		}
+
+		// 5. Redirect user to authorization endpoint.
+		res.writeHead(302, {
+			Location: generateUrl(env.oauth.authorizationEndpoint, locationQuery)
+		});
+		res.end();
+	},
+	logoutMiddleware: function (req, res, next) {
+		var isLogoutPath = req._parsedUrl.pathname === '/logout/'
+		  , accessToken  = res.cookies.get('oAuthToken');
+
+		// 1. If proper path and token is set.
+		if (isLogoutPath && accessToken) {
+			debug('Invalidating token');
+
+			// 1.1. Clean the tokens from cookie.
+			res.cookies.set('oAuthToken', null);
+			res.cookies.set('oAuthRefreshToken', null);
+
+			// 1.2. Invalidate the token in CAS.
+			request({
+				uri: env.oauth.invalidationEndpoint,
+				method: 'GET',
+				headers: {
+					Authorization: 'Bearer ' + accessToken
+				}
+			}, function (error, response, body) {
+				if (error) {
+					debug('Error received from invalidation endpoint:', error);
+				}
+			});
+		}
```

Comment:
I believe it would be safer to put an `else` here (or early return before this point)... in that `else` we would do `next` and the other `next` would be in the body of the callback from request... otherwise it is quite easy to miss sth and make a mistake by mutating sth in the callback and forgetting about handling `next` properly.